specs/electra/beacon-chain.md

```diff
@@ -277,24 +561,353 @@ def get_attesting_indices(state: BeaconState, attestation: Attestation) -> Set[V
     return output
 ```
 
+### Beacon state mutators
+
+#### Updated  `initiate_validator_exit`
+
+```python
+def initiate_validator_exit(state: BeaconState, index: ValidatorIndex) -> None:
+    """
+    Initiate the exit of the validator with index ``index``.
+    """
+    # Return if validator already initiated exit
+    validator = state.validators[index]
+    if validator.exit_epoch != FAR_FUTURE_EPOCH:
+        return
+
+    # Compute exit queue epoch [Modified in Electra:EIP7251]
+    exit_queue_epoch = compute_exit_epoch_and_update_churn(state, validator.effective_balance)
+
+    # Set validator exit epoch and withdrawable epoch
+    validator.exit_epoch = exit_queue_epoch
+    validator.withdrawable_epoch = Epoch(validator.exit_epoch + MIN_VALIDATOR_WITHDRAWABILITY_DELAY)
+```
+
+#### New `switch_to_compounding_validator`
+
+```python
+def switch_to_compounding_validator(state: BeaconState, index: ValidatorIndex) -> None:
+    validator = state.validators[index]
+    if has_eth1_withdrawal_credential(validator):
+        validator.withdrawal_credentials[:1] = COMPOUNDING_WITHDRAWAL_PREFIX
+        queue_excess_active_balance(state, index)
+```
+
+#### New `queue_excess_active_balance`
+
+```python
+def queue_excess_active_balance(state: BeaconState, index: ValidatorIndex) -> None:
+    balance = state.balances[index]
+    if balance > MIN_ACTIVATION_BALANCE:
+        excess_balance = balance - MIN_ACTIVATION_BALANCE
+        state.balances[index] = MIN_ACTIVATION_BALANCE
+        state.pending_balance_deposits.append(
+            PendingBalanceDeposit(index=index, amount=excess_balance)
+        )
+```
+
+#### New `compute_exit_epoch_and_update_churn`
+
+```python
+def compute_exit_epoch_and_update_churn(state: BeaconState, exit_balance: Gwei) -> Epoch:
+    earliest_exit_epoch = compute_activation_exit_epoch(get_current_epoch(state))
+    per_epoch_churn = get_activation_exit_churn_limit(state)
+    # New epoch for exits.
+    if state.earliest_exit_epoch < earliest_exit_epoch:
+        state.earliest_exit_epoch = earliest_exit_epoch
+        state.exit_balance_to_consume = per_epoch_churn
+
+    if exit_balance <= state.exit_balance_to_consume:
+        # Exit fits in the current earliest epoch.
+        state.exit_balance_to_consume -= exit_balance
+    else:
+        # Exit doesn't fit in the current earliest epoch.
+        balance_to_process = exit_balance - state.exit_balance_to_consume
+        additional_epochs, remainder = divmod(balance_to_process, per_epoch_churn)
+        state.earliest_exit_epoch += additional_epochs + 1
```

Comment:
Let's say the current exit epoch is N with churn C. Right now, if you exit balance C, you get still assigned epoch N, and balance to consume is set to 0. If you exit balance 2C, your exit epoch jumps by 2, to N + 2, and the balance to consume is (correctly) set to C. Same if you exit 3C, i.e., exit epoch jumps by 3 to N+3 and balance is set to C. 
This behavior is fine, but it seems to me that one would expect it to not have this jump, and instead behave the same whether you fully consume the remaining churn in the current epoch or up to a future epoch, i.e., if you exit 2C, your exit epoch is set to N with 0 remaining balance to consume. 

```suggestion
        if remainder == 0:
            state.earliest_exit_epoch += additional_epochs
            state.exit_balance_to_consume = 0
        elif remainder > 0:
            state.earliest_exit_epoch += additional_epochs + 1
            state.exit_balance_to_consume = per_epoch_churn - remainder
```

@mkalinin @dapplion I think we discussed this before, let me know if I am forgetting that we made a conscious decision to have this behavior for some reason

Reply:
I think this is rather an off-by-one error that should be fixed. If there is enough capacity for a validator to exit in epoch `N + k` then it should exit in `N + k` and not `N + k + 1`, imho (preserving the status quo behaviour). I was playing around with the fix and came up with the following refactor of the computation:
```python
earliest_exit_epoch = compute_activation_exit_epoch(get_current_epoch(state))
per_epoch_churn = get_activation_exit_churn_limit(state)

if state.earliest_exit_epoch < earliest_exit_epoch:
    exit_balance_to_consume = per_epoch_churn
else:
    exit_balance_to_consume = state.exit_balance_to_consume

# Borrow more capacity if necessary
if exit_balance > exit_balance_to_consume:
    balance_to_process = exit_balance - exit_balance_to_consume
    additional_epochs = (balance_to_process - 1) // per_epoch_churn + 1  # floor(div(x))
    earliest_exit_epoch += additional_epochs
    exit_balance_to_consume += additional_epochs * per_epoch_churn

# Update state variables
state.exit_balance_to_consume = exit_balance_to_consume - exit_balance
state.earliest_exit_epoch = earliest_exit_epoch
```

wdyt?

Reply:
If `state.earliest_exit_epoch > earliest_exit_epoch`, we should ensure that the baseline exit epoch is that one, e.g. `earliest_exit_epoch = min(state.earliest_exit_epoch, compute_activation_exit_epoch(get_current_epoch(state))`

Also `exit_balance_to_consume` here can be more than one epoch worth of churn

Reply:
Nevermind, it looks fine!

Reply:
Checked the behavior more thoroughly with some test cases and everything looks right

Reply:
Checked the behavior more thoroughly with some test cases and everything looks right

Reply:
made an issue to track us fixing this https://github.com/ethereum/consensus-specs/issues/3677